### PR TITLE
Redesign of VLM and LLM blocks

### PR DESCRIPTION
**Pull request by PawelPeczek-Roboflow**
# Description

New ways of using LLMs / VLMs:
* block encapsulating model should let user configure the prompting (giving at least some pre-configured, easy-to-use prompts) - output from that model is string
*  there is group of dedicated formatters intended to parse LLM / VLM output to structured form of desired type

Examples:

**VLM as classifier**
image

<details>
<summary> Workflow definition</summary>

```json
{
  "version": "1.0",
  "inputs": [
    {
      "type": "WorkflowImage",
      "name": "image"
    },
    {
      "type": "WorkflowParameter",
      "name": "open_ai_key"
    },
    {
      "type": "WorkflowParameter",
      "name": "classes"
    }
  ],
  "steps": [
    {
      "type": "roboflow_core/open_ai@v2",
      "name": "gpt",
      "images": "$inputs.image",
      "task_type": "classification",
      "classes": "$inputs.classes",
      "openai_api_key": "$inputs.open_ai_key"
    },
    {
      "type": "roboflow_core/vlm_as_classifier@v1",
      "name": "parser",
      "image": "$inputs.image",
      "vlm_output": "$steps.gpt.output",
      "classes": "$steps.gpt.classes"
    },
    {
      "type": "roboflow_core/property_definition@v1",
      "name": "top_class",
      "operations": [
        {
          "type": "ClassificationPropertyExtract",
          "property_name": "top_class"
        }
      ],
      "data": "$steps.parser.predictions"
    }
  ],
  "outputs": [
    {
      "type": "JsonField",
      "name": "gpt_result",
      "selector": "$steps.gpt.output"
    },
    {
      "type": "JsonField",
      "name": "result",
      "selector": "$steps.top_class.output"
    },
    {
      "type": "JsonField",
      "name": "parsed_prediction",
      "selector": "$steps.parser.*"
    }
  ]
}
```
</details>

❗UI do not render literal as drop-down @hansent / @EmilyGavrilenko - how to change that? I've seen some dropdowns generated in UI already - for instance for Yolo-World version

**Structured prompting**

image

<details>
<summary> Workflow definition</summary>

```json
{
  "version": "1.0",
  "inputs": [
    {
      "type": "WorkflowImage",
      "name": "image"
    },
    {
      "type": "WorkflowParameter",
      "name": "open_ai_key"
    }
  ],
  "steps": [
    {
      "type": "roboflow_core/open_ai@v2",
      "name": "gpt",
      "images": "$inputs.image",
      "task_type": "structured-answering",
      "output_structure": {
        "dogs_count": "count of dogs instances in the image",
        "cats_count": "count of cats instances in the image"
      },
      "openai_api_key": "$inputs.open_ai_key"
    },
    {
      "type": "roboflow_core/json_parser@v1",
      "name": "parser",
      "raw_json": "$steps.gpt.output",
      "expected_fields": [
        "dogs_count",
        "cats_count"
      ]
    },
    {
      "type": "roboflow_core/property_definition@v1",
      "name": "property_definition",
      "operations": [
        {
          "type": "ToString"
        }
      ],
      "data": "$steps.parser.dogs_count"
    }
  ],
  "outputs": [
    {
      "type": "JsonField",
      "name": "result",
      "selector": "$steps.property_definition.output"
    }
  ]
}
```
</details>

❗ The block with JSON parsing rely on dynamic outputs. Outputs can be discovered sending request to server under `POST /workflows/blocks/dynamic_outputs`, but this is not supported on the UI side making the block not useful

## Type of change

Please delete options that are not relevant.

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] This change requires a documentation update

## How has this change been tested, please provide a testcase or example of how you tested the change?

YOUR_ANSWER

## Any specific deployment considerations

For example, documentation changes, usability, usage/costs, secrets, etc.

## Docs

-   [ ] Docs updated? What were the changes:

### CODE REVIEW

File: inference/core/workflows/core_steps/formatters/vlm_as_classifier/v1.py [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+    )
+    classes: Union[
+        WorkflowParameterSelector(kind=[LIST_OF_VALUES_KIND]),
+        StepOutputSelector(kind=[LIST_OF_VALUES_KIND]),
+        List[str],
+    ] = Field(
+        description="List of all classes used by the model, required to "
+        "generate mapping between class name and class id.",
+        examples=[["$steps.lmm.classes", "$inputs.classes", ["class_a", "class_b"]]],
+    )
```
Comment (PawelPeczek-Roboflow) [inline inference/core/workflows/core_steps/formatters/vlm_as_classifier/v1.py:87]: Open question:
* shall we specify dropdown to point model type and task type - I expect that in the future we may need to use that values to choose parsing method specific for given (model, task) pair, yet now it is not needed and would enforce people to choose the option.

Comment (probicheaux) [reply inference/core/workflows/core_steps/formatters/vlm_as_classifier/v1.py:88]: can the model type not be inferred from the previous block?

Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/formatters/vlm_as_classifier/v1.py:88]: not really if we want to fail-fast in compilation, given that there is wrong setup for model/task which is not supported. (compiler do not evaluate steps outputs - hence we would have selector to output not value)

I take the hustle of setting it manually as worth taking if you save on execution costs of API calls only to discover that your workflow is invalid

Comment (probicheaux) [reply inference/core/workflows/core_steps/formatters/vlm_as_classifier/v1.py:88]: Hmm, I am not sure how worfklows works, but can't we have an output type for each llm string, and then make sure that output type is compatible with this block?

Comment (probicheaux) [reply inference/core/workflows/core_steps/formatters/vlm_as_classifier/v1.py:88]: Like the way the blocks are hooked up should be known at compile time so this model_type should be known at compile time, right?

Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/formatters/vlm_as_classifier/v1.py:88]: * output type (kind) for each would solution, but would also be an overkill
* compiler do not understand types of blocks, but kinds of data

File: inference/core/workflows/core_steps/formatters/vlm_as_classifier/v1.py [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+            "license": "Apache-2.0",
+            "block_type": "formatter",
+        }
+    )
+    type: Literal["roboflow_core/vlm_as_classifier@v1"]
+    image: Union[WorkflowImageSelector, StepOutputImageSelector] = Field(
+        description="The image which was the base to generate VLM prediction",
+        examples=["$inputs.image", "$steps.cropping.crops"],
+    )
+    vlm_output: StepOutputSelector(kind=[BATCH_OF_STRING_KIND]) = Field(
```
Comment (probicheaux) [inline inference/core/workflows/core_steps/formatters/vlm_as_classifier/v1.py:75]: new kind here for vlm output string?

**Review by probicheaux — DISMISSED** [review]

File: inference/core/workflows/core_steps/models/foundation/anthropic_claude/v1.py [hunk 3]
```diff
@@ -0,0 +1,657 @@
...
+
+
+def prepare_multi_label_classification_prompt(
+    base64_image: str,
+    classes: List[str],
+    **kwargs,
+) -> Tuple[Optional[str], List[dict]]:
+    serialised_classes = ", ".join(classes)
+    system_prompt = (
+        "You act as multi-label classification model. Your must provide reasonable predictions. "
```
Comment (grzegorz-roboflow) [inline inference/core/workflows/core_steps/models/foundation/anthropic_claude/v1.py:458]:
```suggestion
        "You act as multi-label classification model. You must provide reasonable predictions. "
```

File: inference/core/workflows/core_steps/models/foundation/anthropic_claude/v1.py [hunk 4]
```diff
@@ -0,0 +1,657 @@
...
+        "You are only allowed to produce JSON document. "
+        'Expected structure of json: {"predicted_classes": [{"class": "class-name-1", "confidence": 0.9}, '
+        '{"class": "class-name-2", "confidence": 0.7}]}.'
+        "`class-name-X` must be one of the class name defined by user and `confidence` is a float value "
```
Comment (grzegorz-roboflow) [inline inference/core/workflows/core_steps/models/foundation/anthropic_claude/v1.py:462]:
```suggestion
        "`class-name-X` must be one of the class names defined by user and `confidence` is a float value "
```

File: inference/core/workflows/core_steps/models/foundation/anthropic_claude/v1.py [hunk 5]
```diff
@@ -0,0 +1,657 @@
...
+    return system_prompt, messages
+
+
+def prepare_vqa_prompt(
+    base64_image: str,
+    prompt: str,
+    **kwargs,
+) -> Tuple[Optional[str], List[dict]]:
+    system_prompt = (
+        "You act as Visual Question Answering model. Your task is to provide answer on question"
```
Comment (grzegorz-roboflow) [inline inference/core/workflows/core_steps/models/foundation/anthropic_claude/v1.py:494]:
```suggestion
        "You act as Visual Question Answering model. Your task is to provide answer to question"
```

File: inference/core/workflows/core_steps/models/foundation/anthropic_claude/v1.py [hunk 6]
```diff
@@ -0,0 +1,657 @@
...
+
+def prepare_structured_answering_prompt(
+    base64_image: str,
+    output_structure: Dict[str, str],
+    **kwargs,
+) -> Tuple[Optional[str], List[dict]]:
+    output_structure_serialised = json.dumps(output_structure, indent=4)
+    system_prompt = (
+        f"You are supposed to produce responses in JSON. User is to provide you dictionary with "
+        f"key and values. Each key must be present in your response. Values in user dictionary "
```
Comment (grzegorz-roboflow) [inline inference/core/workflows/core_steps/models/foundation/anthropic_claude/v1.py:585]:
```suggestion
        f"keys and values. Each key must be present in your response. Values in user dictionary "
```

File: inference/core/workflows/core_steps/models/foundation/anthropic_claude/v1.py [hunk 7]
```diff
@@ -0,0 +1,657 @@
...
+
+
+def prepare_object_detection_prompt(
+    base64_image: str,
+    classes: List[str],
+    **kwargs,
+) -> Tuple[Optional[str], List[dict]]:
+    serialised_classes = ", ".join(classes)
+    system_prompt = (
+        "You act as object-detection model. Your must provide reasonable predictions. "
```
Comment (grzegorz-roboflow) [inline inference/core/workflows/core_steps/models/foundation/anthropic_claude/v1.py:618]:
```suggestion
        "You act as object-detection model. You must provide reasonable predictions. "
```

File: inference/core/workflows/core_steps/models/foundation/anthropic_claude/v1.py [hunk 8]
```diff
@@ -0,0 +1,657 @@
...
+        "You are only allowed to produce JSON document. "
+        'Expected structure of json: {"detections": [{"x_min": 0.1, "y_min": 0.2, "x_max": 0.3, "y_max": 0.4, "class_name": "my-class-X", "confidence": 0.7}]} '
+        "- remember to close top-level dictionary at the end. "
+        "`my-class-X` must be one of the class name defined by user. All coordinates must be in range 0.0-1.0, representing percentage of image dimensions. "
```
Comment (grzegorz-roboflow) [inline inference/core/workflows/core_steps/models/foundation/anthropic_claude/v1.py:622]:
```suggestion
        "`my-class-X` must be one of the class names defined by user. All coordinates must be in range 0.0-1.0, representing percentage of image dimensions. "
```

File: inference/core/workflows/core_steps/models/foundation/openai/v2.py [hunk 9]
```diff
@@ -0,0 +1,573 @@
...
+
+
+def prepare_classification_prompt(
+    base64_image: str, classes: List[str], gpt_image_detail: str, **kwargs
+) -> List[dict]:
+    serialised_classes = ", ".join(classes)
+    return [
+        {
+            "role": "system",
+            "content": "You act as single-class classification model. Your must provide reasonable predictions. "
```
Comment (grzegorz-roboflow) [inline inference/core/workflows/core_steps/models/foundation/openai/v2.py:395]:
```suggestion
            "content": "You act as single-class classification model. You must provide reasonable predictions. "
```

File: inference/core/workflows/core_steps/models/foundation/openai/v2.py [hunk 10]
```diff
@@ -0,0 +1,573 @@
...
+) -> List[dict]:
+    serialised_classes = ", ".join(classes)
+    return [
+        {
+            "role": "system",
+            "content": "You act as multi-label classification model. Your must provide reasonable predictions. "
+            "You are only allowed to produce JSON document in Markdown ```json [...]``` markers. "
+            'Expected structure of json: {"predicted_classes": [{"class": "class-name-1", "confidence": 0.9}, '
+            '{"class": "class-name-2", "confidence": 0.7}]}. '
+            "`class-name-X` must be one of the class name defined by user and `confidence` is a float value in range "
```
Comment (grzegorz-roboflow) [inline inference/core/workflows/core_steps/models/foundation/openai/v2.py:431]:
```suggestion
            "`class-name-X` must be one of the class names defined by user and `confidence` is a float value in range "
```

File: inference/core/workflows/core_steps/models/foundation/google_gemini/v1.py [hunk 11]
```diff
@@ -0,0 +1,725 @@
...
+    max_tokens: int,
+    **kwargs,
+) -> dict:
+    serialised_classes = ", ".join(classes)
+    return {
+        "systemInstruction": {
+            "role": "system",
+            "parts": [
+                {
+                    "text": "You act as single-class classification model. Your must provide reasonable predictions. "
```
Comment (grzegorz-roboflow) [inline inference/core/workflows/core_steps/models/foundation/google_gemini/v1.py:410]:
```suggestion
                    "text": "You act as single-class classification model. You must provide reasonable predictions. "
```

File: inference/core/workflows/core_steps/formatters/vlm_as_detector/v1.py [hunk 12]
```diff
@@ -0,0 +1,261 @@
...
+- valid JSON strings
+
+- JSON documents wrapped with Markdown tags
+
+Example
+```
+{"my": "json"}
+```
+
+**Details regarding block behaviour:**
```
Comment (grzegorz-roboflow) [inline inference/core/workflows/core_steps/formatters/vlm_as_detector/v1.py:59]:
```suggestion
**Details regarding block behavior:**
```

File: inference/core/workflows/core_steps/formatters/vlm_as_classifier/v1.py [hunk 13]
```diff
@@ -0,0 +1,269 @@
...
+- valid JSON strings
+
+- JSON documents wrapped with Markdown tags (very common for GPT responses)
+
+Example:
+```
+{"my": "json"}
+```
+
+**Details regarding block behaviour:**
```
Comment (grzegorz-roboflow) [inline inference/core/workflows/core_steps/formatters/vlm_as_classifier/v1.py:47]:
```suggestion
**Details regarding block behavior:**
```

File: inference/core/workflows/core_steps/formatters/json_parser/v1.py [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+Visual Language Models (VLMs). Input is parsed to JSON, and its keys are exposed as block outputs.
+
+Accepted formats:
+- valid JSON strings
+- JSON documents wrapped with Markdown tags (very common for GPT responses)
+```
+{"my": "json"}
+```
+
+**Details regarding block behaviour:**
```
Comment (grzegorz-roboflow) [inline inference/core/workflows/core_steps/formatters/json_parser/v1.py:34]:
```suggestion
**Details regarding block behavior:**
```

File: docs/workflows/kinds.md [hunk 15]
```diff
@@ -23,49 +23,50 @@ for the presence of a mask in the input.
...
-* [`Batch[object_detection_prediction]`](/workflows/kinds/batch_object_detection_prediction): `'predictions'` key from Object Detection Model output
-* [`integer`](/workflows/kinds/integer): Integer value
-* [`roboflow_project`](/workflows/kinds/roboflow_project): Roboflow project name
-* [`Batch[string]`](/workflows/kinds/batch_string): Batch of string values
-* [`image`](/workflows/kinds/image): Image in workflows
-* [`Batch[bar_code_detection]`](/workflows/kinds/batch_bar_code_detection): Prediction with barcode detection
-* [`object_detection_prediction`](/workflows/kinds/object_detection_prediction): Prediction with detected bounding boxes in form of sv.Detections(...) object
-* [`boolean`](/workflows/kinds/boolean): Boolean flag
+* [`roboflow_model_id`](/workflows/kinds/roboflow_model_id): Roboflow model id
+* [`list_of_values`](/workflows/kinds/list_of_values): List of values of any types
```
Comment (grzegorz-roboflow) [inline docs/workflows/kinds.md:61]:
```suggestion
* [`list_of_values`](/workflows/kinds/list_of_values): List of values of any type
```

File: docs/workflows/kinds.md [hunk 16]
```diff
@@ -23,49 +23,50 @@ for the presence of a mask in the input.
...
 !!! Warning
 
-    The list presented below contains elements with `Batch[X]` markers - those will 
-    get soon deprecated and we will use only `X` markers. For now, developers are asked 
-    to create their blocks using the `Batch[X]` markers, but raise the 
-    [issue here](https://github.com/roboflow/inference/issues/608). This GH issue will be used 
-    as a point of communication regarding deprecation process.
+    In `inference` release `0.18.0` we decided to make drastic move to heal the ecosystem 
+    from the problem with ambiguous kinds names (`Batch[X]` vs `X` - see more 
+    [here](https://github.com/roboflow/inference/issues/608)).
```
Comment (grzegorz-roboflow) [inline docs/workflows/kinds.md:28]: I wonder if this should not go into release notes

File: inference/core/workflows/execution_engine/entities/types.py [hunk 17]
```diff
@@ -374,9 +374,75 @@ def __hash__(self) -> int:
...
     class_id=array([2, 7, 2, 0]), 
     tracker_id=None, 
-    data={'class_name': array(['car', 'truck', 'car', 'car'], dtype='<U13')}
+    data={
+        'class_name': array(['car', 'truck', 'car', 'car'], dtype='<U13')
+        'detection_id': array([
+            '51dfa8d5-261c-4dcb-ab30-9aafe9b52379', 'c0c684d1-1e30-4880-aedd-29e67e417264'
+            '8cfc543b-9cfe-493b-b5ad-77afed7bee83', 'c0c684d1-1e30-4880-aedd-38e67e441454'
+        ], dtype='<U36'),
+        'parent_id': array(['image.[0]', 'image.[0]', 'image.[0]', 'image.[0]'], dtype='<U9'),
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/execution_engine/entities/types.py:383]: nope, crop rewrites parents

File: inference/core/workflows/execution_engine/entities/types.py [hunk 18]
```diff
@@ -374,9 +374,75 @@ def __hash__(self) -> int:
...
+bounding boxes and there is no way on how to save metadata for empty predictions). We
+have [GH issue](https://github.com/roboflow/inference/issues/567) to communicate around this
+problem.
+
+**Details of additional fields:**
+
+* `detection_id` - unique identifier for each detection, to be used for when dependent elements 
+are created based on specific detection (example: Dynamic Crop takes this value as parent id for new image)
+
+* `parent_id` - identifier of image that generated prediction (to be fetched from `WorkflowImageData` object)
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/execution_engine/entities/types.py:417]: not really, this comes from the name of input param and the index is batch index, this is opaque btw

File: inference/core/workflows/execution_engine/entities/types.py [hunk 19]
```diff
@@ -374,9 +374,75 @@ def __hash__(self) -> int:
...
+* `inference_id` - identifier of inference request (optional, relevant for Roboflow models)
+
+* `prediction_type` - type of prediction
+
+* `root_parent_id` - identifier of primary Workflow input that was responsible for downstream prediction 
+(to be fetched from `WorkflowImageData` object) - usually identifier of Workflow input placeholder 
+
+* `root_parent_coordinates` - offset regarding origin input - format (`offset_x`, `offset_y`)
+
+* `root_parent_dimensions` - dimensions of origin input image `(height, width)`
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/execution_engine/entities/types.py:430]: not really sure if I understand - we are putting the format of np.arrays here - elements of arrays are 2-elements iterables (height, width) as given

File: inference/core/workflows/core_steps/models/foundation/openai/v2.py [hunk 20]
```diff
@@ -0,0 +1,573 @@
...
+from inference.core.workflows.prototypes.block import (
+    BlockResult,
+    WorkflowBlock,
+    WorkflowBlockManifest,
+)
+
+LONG_DESCRIPTION = """
+Ask a question to OpenAI's GPT-4 with Vision model.
+
+You can specify arbitrary text prompts or predefined ones, the block support the following types of prompt:
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/models/foundation/openai/v2.py:38]: 👍

File: inference/core/workflows/core_steps/models/foundation/google_gemini/v1.py [hunk 21]
```diff
@@ -0,0 +1,725 @@
...
+)
+
+GOOGLE_API_KEY_PATTERN = re.compile(r"key=(.[^&]*)")
+GOOGLE_API_KEY_VALUE_GROUP = 1
+MIN_KEY_LENGTH_TO_REVEAL_PREFIX = 8
+
+LONG_DESCRIPTION = """
+Ask a question to Google's Gemini model with vision capabilities.
+
+You can specify arbitrary text prompts or predefined ones, the block support the following types of prompt:
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/models/foundation/google_gemini/v1.py:43]: 👍

File: inference/core/workflows/core_steps/models/foundation/google_gemini/v1.py [hunk 22]
```diff
@@ -0,0 +1,725 @@
...
+    serialised_classes = ", ".join(classes)
+    return {
+        "systemInstruction": {
+            "role": "system",
+            "parts": [
+                {
+                    "text": "You act as single-class classification model. Your must provide reasonable predictions. "
+                    "You are only allowed to produce JSON document. "
+                    'Expected structure of json: {"class_name": "class-name", "confidence": 0.4}. '
+                    "`class-name` must be one of the class name defined by user. You are only allowed to return "
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/models/foundation/google_gemini/v1.py:413]: 👍

File: inference/core/workflows/core_steps/models/foundation/google_gemini/v1.py [hunk 23]
```diff
@@ -0,0 +1,725 @@
...
+                    "single JSON document, even if there is potentially multiple classes. You are not allowed to "
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/models/foundation/google_gemini/v1.py:414]: 👍

File: inference/core/workflows/core_steps/models/foundation/google_gemini/v1.py [hunk 24]
```diff
@@ -0,0 +1,725 @@
...
+    max_tokens: int,
+    **kwargs,
+) -> dict:
+    serialised_classes = ", ".join(classes)
+    return {
+        "systemInstruction": {
+            "role": "system",
+            "parts": [
+                {
+                    "text": "You act as multi-label classification model. Your must provide reasonable predictions. "
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/models/foundation/google_gemini/v1.py:454]: 👍

File: inference/core/workflows/core_steps/models/foundation/google_gemini/v1.py [hunk 25]
```diff
@@ -0,0 +1,725 @@
...
+                    "You are only allowed to produce JSON document. "
+                    'Expected structure of json: {"predicted_classes": [{"class": "class-name-1", "confidence": 0.9}, '
+                    '{"class": "class-name-2", "confidence": 0.7}]}. '
+                    "`class-name-X` must be one of the class name defined by user and `confidence` is a float value "
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/models/foundation/google_gemini/v1.py:458]: 👍

File: inference/core/workflows/core_steps/models/foundation/google_gemini/v1.py [hunk 26]
```diff
@@ -0,0 +1,725 @@
...
+                    "in range 0.0-1.0 that represent how sure you are that the class is present in the image. "
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/models/foundation/google_gemini/v1.py:459]: 👍

File: inference/core/workflows/core_steps/models/foundation/google_gemini/v1.py [hunk 27]
```diff
@@ -0,0 +1,725 @@
...
+    temperature: Optional[float],
+    max_tokens: int,
+    **kwargs,
+) -> dict:
+    return {
+        "systemInstruction": {
+            "role": "system",
+            "parts": [
+                {
+                    "text": "You act as Visual Question Answering model. Your task is to provide answer on question"
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/models/foundation/google_gemini/v1.py:498]: 👍

File: inference/core/workflows/core_steps/models/foundation/google_gemini/v1.py [hunk 28]
```diff
@@ -0,0 +1,725 @@
...
+    max_tokens: int,
+    **kwargs,
+) -> dict:
+    output_structure_serialised = json.dumps(output_structure, indent=4)
+    return {
+        "systemInstruction": {
+            "role": "system",
+            "parts": [
+                {
+                    "text": f"You are supposed to produce responses in JSON. User is to provide you dictionary with "
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/models/foundation/google_gemini/v1.py:611]: 👍

File: inference/core/workflows/core_steps/models/foundation/google_gemini/v1.py [hunk 29]
```diff
@@ -0,0 +1,725 @@
...
+                    f"key and values. Each key must be present in your response. Values in user dictionary "
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/models/foundation/google_gemini/v1.py:612]: 👍

File: inference/core/workflows/core_steps/models/foundation/google_gemini/v1.py [hunk 30]
```diff
@@ -0,0 +1,725 @@
...
+                    f"represent descriptions for JSON fields to be generated. Provide only JSON in response.",
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/models/foundation/google_gemini/v1.py:613]: 👍

File: inference/core/workflows/core_steps/models/foundation/openai/v2.py [hunk 31]
```diff
@@ -0,0 +1,573 @@
...
+    base64_image: str, classes: List[str], gpt_image_detail: str, **kwargs
+) -> List[dict]:
+    serialised_classes = ", ".join(classes)
+    return [
+        {
+            "role": "system",
+            "content": "You act as single-class classification model. Your must provide reasonable predictions. "
+            "You are only allowed to produce JSON document in Markdown ```json [...]``` markers. "
+            'Expected structure of json: {"class_name": "class-name", "confidence": 0.4}. '
+            "`class-name` must be one of the class name defined by user. You are only allowed to return "
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/models/foundation/openai/v2.py:398]: 👍

File: inference/core/workflows/core_steps/models/foundation/openai/v2.py [hunk 32]
```diff
@@ -0,0 +1,573 @@
...
+            "single JSON document, even if there is potentially multiple classes. You are not allowed to return list.",
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/models/foundation/openai/v2.py:399]: 👍

File: inference/core/workflows/core_steps/models/foundation/openai/v2.py [hunk 33]
```diff
@@ -0,0 +1,573 @@
...
+
+
+def prepare_multi_label_classification_prompt(
+    base64_image: str, classes: List[str], gpt_image_detail: str, **kwargs
+) -> List[dict]:
+    serialised_classes = ", ".join(classes)
+    return [
+        {
+            "role": "system",
+            "content": "You act as multi-label classification model. Your must provide reasonable predictions. "
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/models/foundation/openai/v2.py:427]: 👍

File: inference/core/workflows/core_steps/models/foundation/openai/v2.py [hunk 34]
```diff
@@ -0,0 +1,573 @@
...
+    ]
+
+
+def prepare_vqa_prompt(
+    base64_image: str, prompt: str, gpt_image_detail: str, **kwargs
+) -> List[dict]:
+    return [
+        {
+            "role": "system",
+            "content": "You act as Visual Question Answering model. Your task is to provide answer on question"
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/models/foundation/openai/v2.py:460]: 👍

File: inference/core/workflows/core_steps/models/foundation/openai/v2.py [hunk 35]
```diff
@@ -0,0 +1,573 @@
...
+
+def prepare_structured_answering_prompt(
+    base64_image: str, output_structure: Dict[str, str], gpt_image_detail: str, **kwargs
+) -> List[dict]:
+    output_structure_serialised = json.dumps(output_structure, indent=4)
+    return [
+        {
+            "role": "system",
+            "content": f"You are supposed to produce responses in JSON wrapped in Markdown markers: "
+            f"```json\nyour-response\n```. User is to provide you dictionary with key and values. "
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/models/foundation/openai/v2.py:540]: 👍

File: inference/core/workflows/core_steps/models/foundation/anthropic_claude/v1.py [hunk 36]
```diff
@@ -0,0 +1,657 @@
...
+from inference.core.workflows.prototypes.block import (
+    BlockResult,
+    WorkflowBlock,
+    WorkflowBlockManifest,
+)
+
+LONG_DESCRIPTION = """
+Ask a question to Anthropic Claude model with vision capabilities.
+
+You can specify arbitrary text prompts or predefined ones, the block support the following types of prompt:
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/models/foundation/anthropic_claude/v1.py:41]: 👍

File: inference/core/workflows/core_steps/models/foundation/anthropic_claude/v1.py [hunk 37]
```diff
@@ -0,0 +1,657 @@
...
+
+- `unconstrained` - any arbitrary prompt you like 
+
+- `ocr`- predefined prompt to recognise text from image
+
+- `visual-question-answering` - your prompt is supposed to provide question and will be
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/models/foundation/anthropic_claude/v1.py:47]: 👎

File: inference/core/workflows/core_steps/models/foundation/anthropic_claude/v1.py [hunk 38]
```diff
@@ -0,0 +1,657 @@
...
+    base64_image: str,
+    classes: List[str],
+    **kwargs,
+) -> Tuple[Optional[str], List[dict]]:
+    serialised_classes = ", ".join(classes)
+    system_prompt = (
+        "You act as single-class classification model. Your must provide reasonable predictions. "
+        "You are only allowed to produce JSON document. "
+        'Expected structure of json: {"class_name": "class-name", "confidence": 0.4}. '
+        "`class-name` must be one of the class name defined by user. You are only allowed to return "
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/models/foundation/anthropic_claude/v1.py:425]: 👍

File: inference/core/workflows/core_steps/models/foundation/anthropic_claude/v1.py (hunk 3 above)
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/models/foundation/anthropic_claude/v1.py:458]: 👍

File: inference/core/workflows/core_steps/models/foundation/anthropic_claude/v1.py (hunk 4 above)
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/models/foundation/anthropic_claude/v1.py:462]: 👍

File: inference/core/workflows/core_steps/models/foundation/anthropic_claude/v1.py (hunk 5 above)
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/models/foundation/anthropic_claude/v1.py:494]: 👍

File: inference/core/workflows/core_steps/models/foundation/anthropic_claude/v1.py (hunk 6 above)
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/models/foundation/anthropic_claude/v1.py:585]: 👍

File: inference/core/workflows/core_steps/models/foundation/anthropic_claude/v1.py (hunk 7 above)
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/models/foundation/anthropic_claude/v1.py:618]: 👍

File: inference/core/workflows/core_steps/models/foundation/anthropic_claude/v1.py (hunk 8 above)
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/models/foundation/anthropic_claude/v1.py:622]: 👍

File: inference/core/workflows/core_steps/models/foundation/openai/v2.py (hunk 9 above)
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/models/foundation/openai/v2.py:395]: 👍

File: inference/core/workflows/core_steps/models/foundation/openai/v2.py (hunk 10 above)
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/models/foundation/openai/v2.py:431]: 👍

File: inference/core/workflows/core_steps/models/foundation/google_gemini/v1.py (hunk 11 above)
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/models/foundation/google_gemini/v1.py:410]: 👍

File: inference/core/workflows/core_steps/formatters/vlm_as_classifier/v1.py (hunk 13 above)
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/formatters/vlm_as_classifier/v1.py:47]: 👍

File: inference/core/workflows/core_steps/formatters/json_parser/v1.py (hunk 14 above)
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/formatters/json_parser/v1.py:34]: 👍

File: docs/workflows/kinds.md (hunk 15 above)
Comment (PawelPeczek-Roboflow) [reply docs/workflows/kinds.md:61]: 👍

File: docs/workflows/kinds.md (hunk 16 above)
Comment (PawelPeczek-Roboflow) [reply docs/workflows/kinds.md:28]: would be there

**Review by grzegorz-roboflow — APPROVED** [review]

**Pull request merged**